include/secp256k1_schnorrsig.h

```diff
@@ -56,6 +56,85 @@ SECP256K1_API int secp256k1_schnorrsig_parse(
+ *
+ *  The next functions can be used to prevent a signing device from exfiltrating the secret signing
+ *  keys through biased signature nonces. The general idea is that a host provides additional
+ *  randomness to the signing device client and the client commits to the randomness in the nonce
+ *  using sign-to-contract.
+ *  In order to make the randomness unpredictable, the host and client must engage in a
+ *  commit-reveal protocol as follows:
+ *  1. The host draws the randomness, commits to it with the `anti_nonce_sidechan_host_commit`
+ *     function and sends the commitment to the client.
+ *  2. The client commits to its sign-to-contract original nonce (which is the nonce without the
+ *     sign-to-contract tweak) using the hosts commitment by calling the
+ *     `secp256k1_schnorrsig_anti_nonce_sidechan_client_commit` function. The client sends the
+ *     rusulting commitment to the host
+ *  3. The host replies with the randomness generated in step 1.
+ *  4. The client signs with `schnorrsig_sign` using the host provided randomness as `s2c_data` and
+ *     sends the signature and opening to the host.
+ *  5. The host checks that the signature contains an sign-to-contract commitment to the randomness
+ *     by calling `secp256k1_schnorrsig_anti_nonce_sidechan_host_verify` with the client's
+ *     commitment from step 2 and the signature and opening received in step 4. If verification does
+ *     not succeed, the protocol failed and can be restarted.
+ */
+
+/** Create a randomness commitment on the host as part of the Anti Nonce Sidechannel Protocol.
+ *
+ *  Returns 1 on success, 0 on failure.
+ *  Args:              ctx: pointer to a context object (cannot be NULL)
+ *  Out: rand_commitment32: pointer to 32-byte array to store the returned commitment (cannot be NULL)
+ *  In:             rand32: the 32-byte randomness to commit to (cannot be NULL)
+ */
+SECP256K1_API int secp256k1_schnorrsig_anti_nonce_sidechan_host_commit(
+    secp256k1_context *ctx,
+    unsigned char *rand_commitment32,
+    const unsigned char *rand32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3);
+
+/** Compute commitment on the client as part of the Anti Nonce Sidechannel Protocol.
+ *
+ *  Returns 1 on success, 0 on failure.
+ *  Args:           ctx: pointer to a context object (cannot be NULL)
+ *  Out:  client_commit: pointer to a pubkey where the clients public nonce will be
+ *                       placed. This is the public nonce before doing the
+ *                       sign-to-contract commitment to the hosts randomness (cannot
+ *                       be NULL)
+ *  In:           msg32: the 32-byte message hash to be signed (cannot be NULL)
+ *             seckey32: the 32-byte secret key used for signing (cannot be NULL)
+ *    rand_commitment32: the 32-byte randomness commitment from the host (cannot be NULL)
+ */
+SECP256K1_API int secp256k1_schnorrsig_anti_nonce_sidechan_client_commit(
```

Comment:
To keep the api / interface smaller, the client could also call the sign function and get the commitment from the signature. Would there be any security downside in that (except for code readability?).

Reply:
That doesn't seem to work if I understand you correctly. The client needs the hosts randomness to produce the signature which it doesn't get until it sends the commitment.

Reply:
I meant a first of two calls to the signing function, with the first one being a plain call with no host randomness, just to extract the client commitment (`R = k*G`, `R` can be reconstructed using the sig).

Reply:
If you call `sign` twice your wasting a lot of time producing a signature that's unnecessary?

Reply:
:+1: I see.